src/core/algorithms/nd/nd_verifier/nd_verifier.cpp

```diff
@@ -0,0 +1,234 @@
+                          config::names::kWeight});
+}
+
+unsigned long long NDVerifier::ExecuteInternal() {
+    LOG(INFO) << "Parameters of NDVerifier:";
+    LOG(INFO) << "\tInput table: " << input_table_->GetRelationName();
+    LOG(INFO) << "\tNull equals null: " << is_null_equal_null_;
+    LOG(INFO) << "\tLhs indices: " << util::VectorToString(lhs_indices_);
+    LOG(INFO) << "\tRhs indices: " << util::VectorToString(rhs_indices_);
+    LOG(INFO) << "\tWeight: " << weight_;
+
+    auto verification_time = ::util::TimedInvoke(&NDVerifier::VerifyND, this);
+
+    LOG(DEBUG) << "ND verification took " << std::to_string(verification_time) << "ms";
+
+    auto stats_calculation_time = ::util::TimedInvoke(&NDVerifier::CalculateStats, this);
+
+    LOG(DEBUG) << "Statistics calculation took " << std::to_string(stats_calculation_time) << "ms";
+
+    return verification_time + stats_calculation_time;
+}
+
+bool NDVerifier::NDHolds() const {
+    return stats_calculator_.GetRealWeight() <= weight_;
+}
+
+void NDVerifier::ResetState() {
+    stats_calculator_ = util::StatsCalculator{};
+}
+
+NDVerifier::EncodedValuesType NDVerifier::EncodeMultipleValues(
+        config::IndicesType const& col_idxs) const {
+    auto values = std::make_shared<std::vector<std::string>>();
+    auto row = std::make_shared<std::vector<size_t>>();
+
+    for (size_t row_idx{0}; row_idx < typed_relation_->GetNumRows(); ++row_idx) {
+        std::stringstream ss;
+        ss << '(';
+        bool is_null = false;
+        for (auto col_idx_pt{col_idxs.begin()}; col_idx_pt != col_idxs.end(); ++col_idx_pt) {
+            model::TypedColumnData const& col_data = typed_relation_->GetColumnData(*col_idx_pt);
+            std::vector<std::byte const*> const& byte_data = col_data.GetData();
+            model::Type const& type = col_data.GetType();
+
+            if (col_idx_pt != col_idxs.begin()) {
+                ss << ", ";
+            }
+
+            std::byte const* bytes_ptr = byte_data[row_idx];
+            if (bytes_ptr == nullptr) {
+                LOG(INFO) << "WARNING: Cell (" << *col_idx_pt << ", " << row_idx << ") is empty";
+                is_null = true;
+            } else {
+                ss << type.ValueToString(bytes_ptr);
+            }
+        }
+        ss << ')';
+
+        auto string_data = ss.str();
+        size_t index{values->size()};
+        if (is_null && !is_null_equal_null_) {
+            // If not null_eq_null, every value with null will be unique
+            values->push_back(string_data);
+        } else {
+            for (size_t i{0}; i < values->size(); ++i) {
+                if ((*values)[i] == string_data) {
+                    index = i;
+                    break;
+                }
+            }
+
+            if (index == values->size()) {  // wasn't in codes
+                values->push_back(string_data);
+            }
+        }
+        row->push_back(index);
```

Comment:
I see duplication of this code. Can we separate big loop into two parts (move each part into a separate method) and then use one of these methods here and in `EncodeSingleValues()` method?